### DISCUSSION TITLE
Getting "ImportError: cannot import name 'ComposeResult' from 'textual.app'"

### DISCUSSION THREAD

**Discussion by MJohnson3150** [Textualize/textual#1452] [Q&A] [answered]
Very new to Textual and Python in general. I’m trying to run and play with some of the examples. Several have the following import line:

from textual.app import App, ComposeResult

When I try running this line, I get the following error:

ImportError: cannot import name ‘ComposeResult’ from ‘textual.app’

In looking at another post, it was recommended to run:
python3 -m textual 
to see a list of its capabilities. I'm also not able to run this line. In this case, I get the following:
No module named textual.__main__; 'textual' is a package and cannot be directly executed

I have installed the Textual library successfully and noticed that I can import things from textual but not textual.app. Is that something else? Sorry for the cluelessness but I don’t know enough yet on how to find out for myself. (I've tried Googling several things but I must be missing something very fundamental that I can't find. Thank you.

**Comment by willmcgugan** [comment]
Did you create a filled called “textual.py”? This will break things. The fix is to rename it to something else.

**Reply by MJohnson3150** [reply]
Oh man! I was excited by this because I thought I may have done this but the only file I created like this was textual1.py. I went ahead and deleted that just in case but I still get the same error. I don't have a textual.py in any of my directories.

**Reply by willmcgugan** [reply]
What happens if you do this?

```python
python -c "import textual; print(textual)"
```

**Reply by MJohnson3150** [reply]
I get the following:
mjohnson@ubuntu20:~/Documents/Scripts$ python3 -c "import textual; print(textual)"
<module 'textual' from '/home/mjohnson/.local/lib/python3.8/site-packages/textual/__init__.py'>

Thank you, Will.

**Comment by MJohnson3150** [comment]
Hi again and Happy New Year.

Any idea on this? Thanks.

Marc Johnson  |  Senior Manager of IT Systems
Las Vegas Convention and Visitors Authority  |  LVCVA.com
T: 702-892-0711 | Direct: 702-892-7601

**Reply by davep** [reply]
Could you run this in your Python REPL please?

```python
from importlib.metadata import version; version( "textual" )
```

and say what you get back?

The reason I ask is that `ComposeResult` wasn't introduced until the merge of the "CSS branch", resulting in [v0.2.0 being released back in October last year](https://github.com/Textualize/textual/releases/tag/v0.2.0). If I do what you're doing, but with [v0.1.18 or earlier](https://github.com/Textualize/textual/releases/tag/v0.1.18), I get the same result as you. My suspicion is that you're looking at our current documentation, which is for the current versions of Textual, but you've got an older version installed.

The current stable version is [v0.9.1](https://pypi.org/project/textual/).

**Comment by MJohnson3150** [comment]
Thanks for responding Dave.

Yes, I get back: ‘0.1.18’

So I thought I would try to update the textual library with:

pip install "textual[dev]"

and I get the following:
pip3 install "textual[dev]"
Requirement already satisfied: textual[dev] in /home/mjohnson/.local/lib/python3.8/site-packages (0.1.18)
  WARNING: textual 0.1.18 does not provide the extra 'dev'
Requirement already satisfied: rich<13.0.0,>=12.3.0 in /home/mjohnson/.local/lib/python3.8/site-packages (from textual[dev]) (12.5.1)
Requirement already satisfied: typing-extensions<5.0,>=4.0.0; python_version < "3.9" in /home/mjohnson/.local/lib/python3.8/site-packages (from textual[dev]) (4.2.0)
Requirement already satisfied: pygments<3.0.0,>=2.6.0 in /home/mjohnson/.local/lib/python3.8/site-packages (from rich<13.0.0,>=12.3.0->textual[dev]) (2.11.2)
Requirement already satisfied: commonmark<0.10.0,>=0.9.0 in /home/mjohnson/.local/lib/python3.8/site-packages (from rich<13.0.0,>=12.3.0->textual[dev]) (0.9.1)

And if I run your command again, it still says I’m on the 0.1.18 version. So how do I get it to the latest? Thanks again.

Marc Johnson  |  Senior Manager of IT Systems
Las Vegas Convention and Visitors Authority  |  LVCVA.com
T: 702-892-0711 | Direct: 702-892-7601

**Reply by willmcgugan** [reply]
Add  `-U` switch to upgrade to the latest version.

**Comment by MJohnson3150** [comment, marked as answer]
That did it! Thanks again for all the help. I look forward to learning Textual.

Marc Johnson  |  Senior Manager of IT Systems
Las Vegas Convention and Visitors Authority  |  LVCVA.com
T: 702-892-0711 | Direct: 702-892-7601

**Reply by davep** [reply]
Glad to hear you've got it sorted. Any other questions, feel free to ask here [or over on the Discord server](https://discord.gg/Enf6Z3qhVr).
